File: src/main/java/dev/strawn/frugal/FrugalSyntaxHighlighterFactory.java
package dev.strawn.frugal;

import com.intellij.openapi.fileTypes.SyntaxHighlighterFactory;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors;
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase;
import com.intellij.lexer.Lexer;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class FrugalSyntaxHighlighterFactory extends SyntaxHighlighterFactory {
    
    @Override
    public @NotNull SyntaxHighlighterBase getSyntaxHighlighter(@Nullable Project project, @Nullable VirtualFile virtualFile) {
        return new FrugalSyntaxHighlighter();
    }
    
    public static class FrugalSyntaxHighlighter extends SyntaxHighlighterBase {
        
        public static final TextAttributesKey KEYWORD = TextAttributesKey.createTextAttributesKey(
            "FRUGAL_KEYWORD",
            DefaultLanguageHighlighterColors.KEYWORD
        );
        
        public static final TextAttributesKey STRING = TextAttributesKey.createTextAttributesKey(
            "FRUGAL_STRING", 
            DefaultLanguageHighlighterColors.STRING
        );
        
        public static final TextAttributesKey NUMBER = TextAttributesKey.createTextAttributesKey(
            "FRUGAL_NUMBER",
            DefaultLanguageHighlighterColors.NUMBER
        );
        
        public static final TextAttributesKey COMMENT = TextAttributesKey.createTextAttributesKey(
            "FRUGAL_COMMENT",
            DefaultLanguageHighlighterColors.LINE_COMMENT
        );
        
        public static final TextAttributesKey IDENTIFIER = TextAttributesKey.createTextAttributesKey(
            "FRUGAL_IDENTIFIER",
            DefaultLanguageHighlighterColors.IDENTIFIER
        );
        
        private static final TextAttributesKey[] KEYWORDS = new TextAttributesKey[]{KEYWORD};
        private static final TextAttributesKey[] STRINGS = new TextAttributesKey[]{STRING};
        private static final TextAttributesKey[] NUMBERS = new TextAttributesKey[]{NUMBER};
        private static final TextAttributesKey[] COMMENTS = new TextAttributesKey[]{COMMENT};
        private static final TextAttributesKey[] IDENTIFIERS = new TextAttributesKey[]{IDENTIFIER};
        private static final TextAttributesKey[] EMPTY = new TextAttributesKey[0];
        
        @Override
        public @NotNull Lexer getHighlightingLexer() {
            return new FrugalLexer();
        }
        
        @Override
        public TextAttributesKey @NotNull [] getTokenHighlights(IElementType tokenType) {
            if (tokenType == FrugalTokenTypes.NAMESPACE ||
                tokenType == FrugalTokenTypes.INCLUDE ||
                tokenType == FrugalTokenTypes.SERVICE ||
                tokenType == FrugalTokenTypes.STRUCT ||
                tokenType == FrugalTokenTypes.ENUM ||
                tokenType == FrugalTokenTypes.EXCEPTION ||
                tokenType == FrugalTokenTypes.UNION ||
                tokenType == FrugalTokenTypes.TYPEDEF) {
                return KEYWORDS;
            } else if (tokenType == FrugalTokenTypes.STRING) {
                return STRINGS;
            } else if (tokenType == FrugalTokenTypes.NUMBER) {
                return NUMBERS;
            } else if (tokenType == FrugalTokenTypes.COMMENT) {
                return COMMENTS;
            } else if (tokenType == FrugalTokenTypes.IDENTIFIER) {
                return IDENTIFIERS;
            } else {
                return EMPTY;
            }
        }
    }
}

File: src/main/kotlin/dev/strawn/frugal/FrugalSyntaxHighlighterFactory.kt
package dev.strawn.frugal

import com.intellij.openapi.fileTypes.SyntaxHighlighterFactory
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase
import com.intellij.lexer.Lexer
import com.intellij.psi.tree.IElementType

class FrugalSyntaxHighlighterFactory : SyntaxHighlighterFactory() {
    
    override fun getSyntaxHighlighter(project: Project?, virtualFile: VirtualFile?): FrugalSyntaxHighlighter {
        return FrugalSyntaxHighlighter()
    }
}

class FrugalSyntaxHighlighter : SyntaxHighlighterBase() {
    
    companion object {
        val KEYWORD = TextAttributesKey.createTextAttributesKey(
            "FRUGAL_KEYWORD",
            DefaultLanguageHighlighterColors.KEYWORD
        )
        
        val STRING = TextAttributesKey.createTextAttributesKey(
            "FRUGAL_STRING", 
            DefaultLanguageHighlighterColors.STRING
        )
        
        val NUMBER = TextAttributesKey.createTextAttributesKey(
            "FRUGAL_NUMBER",
            DefaultLanguageHighlighterColors.NUMBER
        )
        
        val COMMENT = TextAttributesKey.createTextAttributesKey(
            "FRUGAL_COMMENT",
            DefaultLanguageHighlighterColors.LINE_COMMENT
        )
        
        val IDENTIFIER = TextAttributesKey.createTextAttributesKey(
            "FRUGAL_IDENTIFIER",
            DefaultLanguageHighlighterColors.IDENTIFIER
        )
        
        private val KEYWORDS = arrayOf(KEYWORD)
        private val STRINGS = arrayOf(STRING)
        private val NUMBERS = arrayOf(NUMBER)
        private val COMMENTS = arrayOf(COMMENT)
        private val IDENTIFIERS = arrayOf(IDENTIFIER)
        private val EMPTY = emptyArray<TextAttributesKey>()
    }
    
    override fun getHighlightingLexer(): Lexer {
        return FrugalLexer()
    }
    
    override fun getTokenHighlights(tokenType: IElementType?): Array<TextAttributesKey> {
        return when (tokenType) {
            FrugalTokenTypes.NAMESPACE,
            FrugalTokenTypes.INCLUDE,
            FrugalTokenTypes.SERVICE,
            FrugalTokenTypes.STRUCT,
            FrugalTokenTypes.ENUM,
            FrugalTokenTypes.EXCEPTION,
            FrugalTokenTypes.UNION,
            FrugalTokenTypes.TYPEDEF -> KEYWORDS
            
            FrugalTokenTypes.STRING -> STRINGS
            FrugalTokenTypes.NUMBER -> NUMBERS
            FrugalTokenTypes.COMMENT -> COMMENTS
            FrugalTokenTypes.IDENTIFIER -> IDENTIFIERS
            else -> EMPTY
        }
    }
}